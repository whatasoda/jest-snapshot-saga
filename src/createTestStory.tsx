import React, { ComponentType, Dispatch, useState, useEffect, FunctionComponent } from 'react';
import { render, act } from '@testing-library/react';
import { ROOT_ELEMENT, MONOLITH_SNAPSHOT } from './utils/symbols';
import {
  StoryOptions,
  TestStoryType,
  TestStorySnapshotState,
  SnapshotSectionRecord,
  MonolithSnapshotState,
} from './decls';
import injectRenderCounter from './utils/injectRenderCounter';
import serialize from './serialize';
import pseudoEventTarget from './mock/pseudoEventTarget';

const start = <P extends object>(Component: ComponentType<P>, { Provider, functions }: StoryOptions = {}) => {
  const { Injected, getRenderCount } = injectRenderCounter(Component);
  let dispatchProps = (null as unknown) as Dispatch<P | null>;

  const StoryComponent: FunctionComponent = () => {
    const [props, setProps] = useState<P | null>(null);
    useEffect(() => {
      dispatchProps = setProps;
    }, []);

    if (!props) return null;
    return <Injected {...props} />;
  };

  const setProps = (props: P) => {
    act(() => dispatchProps(props));
  };

  const unmount = () => {
    act(() => dispatchProps(null));
  };

  const result = render(<StoryComponent />, { wrapper: Provider });
  const snapshotState: TestStorySnapshotState = {
    [ROOT_ELEMENT]: true,
    root: result.container,
    isMonolith: false,
    getRenderCount,
    functions: functions ? functions.filter(jest.isMockFunction) : [],
    prev: { elements: '', functions: '', styles: '' },
    diff: { elements: false, functions: false, styles: false },
  };

  return { result, getRenderCount, setProps, unmount, snapshotState };
};

const createTestStory = <P extends object>(
  Component: ComponentType<P>,
  options: StoryOptions = {},
): (() => TestStoryType<P>) => {
  return () => {
    const { snapshotState, ...general } = start(Component, options);

    const snapshot = expect(snapshotState).toMatchSnapshot;
    const setDiffState = (diff: Partial<SnapshotSectionRecord<boolean>> = {}) => {
      Object.assign(snapshotState.diff, diff);
    };

    return { ...pseudoEventTarget, ...general, snapshot, setDiffState };
  };
};

createTestStory.monolith = <P extends object>(
  Component: ComponentType<P>,
  options: StoryOptions = {},
): (() => TestStoryType<P> & { finish: () => void }) => {
  return () => {
    const { snapshotState, ...general } = start(Component, options);
    snapshotState.isMonolith = true;

    const monolithSnapshotState: MonolithSnapshotState = {
      [MONOLITH_SNAPSHOT]: true,
      list: [],
    };

    const snapshot = (description?: string) => {
      monolithSnapshotState.list.push(serialize({ ...snapshotState, description }));
      return snapshotState;
    };
    const setDiffState = (diff: Partial<SnapshotSectionRecord<boolean>> = {}) => {
      Object.assign(snapshotState.diff, diff);
    };

    const finish = expect(monolithSnapshotState).toMatchSnapshot;

    return { ...pseudoEventTarget, ...general, snapshot, setDiffState, finish };
  };
};

export default createTestStory;
