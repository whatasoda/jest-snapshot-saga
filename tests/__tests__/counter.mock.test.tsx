import createTestStory, { createMockComponent } from '../../src';

const Counter = createMockComponent<typeof import('../components/functional/Counter').default>('Counter');
jest.mock('../components/functional/Counter.tsx', () => Counter);

import CounterContainer from '../containers/functional/CounterContainer';

describe('Counter splitted with mock component', () => {
  const setDozen = jest.fn().mockName('setDozen');

  const start = createTestStory(CounterContainer, {
    extraMockFunctions: [setDozen],
  });

  it('render', () => {
    const story = start();
    story.snapshot();

    story.setProps({ setDozen });
    story.snapshot();
  });
});

describe('Counter with mock component', () => {
  Counter.mockEffect(({ increment }, { setEventListener, order }) => {
    setEventListener(`${order}: increment`, increment);
  });
  const setDozen = jest.fn().mockName('setDozen');

  const start = createTestStory.monolith(CounterContainer, {
    extraMockFunctions: [setDozen],
  });

  it('render', () => {
    const story = start();
    story.snapshot();

    story.setProps({ setDozen });
    story.snapshot('"setDozen" should be called once');

    Array.from({ length: 11 }).forEach(() => {
      story.dispatch('0: increment');
    });
    story.snapshot('"setDozen" must not be called');

    story.dispatch('0: increment');
    story.snapshot('"setDozen" should be called once more');

    story.finish();
  });
});
