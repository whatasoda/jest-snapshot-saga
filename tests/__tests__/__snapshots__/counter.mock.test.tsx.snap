// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter splitted with mock component render 1`] = `
  render call: 0 in total
  >>>>>> functions >>>>>>>>>>>>>>>
    /* none */
  >>>>>> styles >>>>>>>>>>>>>>>>>>
    /* none */
  >>>>>> elements >>>>>>>>>>>>>>>>
    /* none */
`;

exports[`Counter splitted with mock component render 2`] = `
  render call: 1 in total
  >>>>>> functions >>>>>>>>>>>>>>>

    setDozen /* #0 */ (
      0
    ) /* return */ => (undefined)

  >>>>>> styles >>>>>>>>>>>>>>>>>>
    /* none */
  >>>>>> elements >>>>>>>>>>>>>>>>

    <ROOT>
      <MockComponent Counter
        length={0}
        increment={[Function]}
      />
    </ROOT>
`;

exports[`Counter with mock component render 1`] = `
<< 0 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

  render call: 0 in total
  >>>>>> functions >>>>>>>>>>>>>>>
    /* none */
  >>>>>> styles >>>>>>>>>>>>>>>>>>
    /* none */
  >>>>>> elements >>>>>>>>>>>>>>>>
    /* none */

<< 1 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    'setDozen' should be called once
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  render call: 1 in total
  >>>>>> functions >>>>>>>>>>>>>>>

    setDozen /* #0 */ (
      0
    ) /* return */ => (undefined)

  >>>>>> styles >>>>>>>>>>>>>>>>>>
    /* none */
  >>>>>> elements >>>>>>>>>>>>>>>>

    <ROOT>
      <MockComponent Counter
        length={0}
        increment={[Function]}
      />
    </ROOT>

<< 2 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    'setDozen must not be called
    'length' of Conunter should be 11
    render call should be 12
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  render call: 12 in total
  >>>>>> functions >>>>>>>>>>>>>>>
    /* none */
  >>>>>> styles >>>>>>>>>>>>>>>>>>
    /* none */
  >>>>>> elements >>>>>>>>>>>>>>>>

    <ROOT>
      <MockComponent Counter
        length={11}
        increment={[Function]}
      />
    </ROOT>

<< 3 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    'setDozen' should be called once more
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  render call: 13 in total
  >>>>>> functions >>>>>>>>>>>>>>>

    setDozen /* #0 */ (
      1
    ) /* return */ => (undefined)

  >>>>>> styles >>>>>>>>>>>>>>>>>>
    /* none */
  >>>>>> elements >>>>>>>>>>>>>>>>

    <ROOT>
      <MockComponent Counter
        length={12}
        increment={[Function]}
      />
    </ROOT>

<< END <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
`;
